### PR TITLE
[Translation] Remove TranslatorBagInterface to allow for optimized caching in 3.0

### PR DESCRIPTION
**Pull request by mpdude**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR | n/a |

We would like the `Translator` to be able to perform various optimizations on the `MessageCatalogues` it holds internally. One particular example is working with reduced catalogues for fallback locales that only contain the messages not present in one of the more preferred locales, or possibly do a lazy-loading of catalogues for fallback locales.

These optimizations are not possible as long as clients are able to retrieve the `MessageCatalogue` instances used inside the `Translator` by means of the `TranslatorBagInterface` and expect them to be full-fledged and complete catalogue instances (so the initial attempt had to be reverted in #14315, also see #14380 for a discussion). 

In other words, `TranslatorBagInterface` breaks a boundary of encapsulation important for optimizations.

The two clients of the interface inside Symfony are not really interested in the catalogues per se, but only need them to iterate along the fallback chain in order to figure out the locale ultimately used. This information can now be obtained through the new `FallbackLocaleAwareInterface`.

**To be discussed:**
- In 3.0, should `FallbackLocaleAwareInterface` remain a dedicated interface or become part of `TranslatorInterface`?
- Shall we keep the `TranslatorBagInterface` in the future, especially if `Translator` does not implement it?

### CODE REVIEW

File: src/Symfony/Component/Translation/Translator.php [hunk 1]
```diff
@@ -506,4 +522,19 @@ private function getConfigCacheFactory()
 
         return $this->configCacheFactory;
     }
+
+    private function getCatalogueInternal($locale = null)
```
Comment (mpdude) [inline src/Symfony/Component/Translation/Translator.php]: We could rename this back to `getCatalogue()` and make it private once we do no longer implement `TranslatorBagInterface` (in 3.0).

File: src/Symfony/Component/Translation/Translator.php [hunk 2]
```diff
@@ -241,36 +241,52 @@ public function transChoice($id, $number, array $parameters = array(), $domain =
         }
 
         $id = (string) $id;
-        $catalogue = $this->getCatalogue($locale);
+        $bestLocale = $this->resolveLocale($id, $domain, $locale);
+
+        if (null !== $bestLocale) {
+           return strtr($this->selector->choose($this->getCatalogueInternal($locale)->get($id, $domain), (int) $number, $bestLocale), $parameters);
+        } else {
+           return strtr($this->selector->choose($id, (int) $number, $locale), $parameters);
```
Comment (mpdude) [inline src/Symfony/Component/Translation/Translator.php]: Might be an edge case not covered by tests – which locale should the selector apply when the message has no translation, i. e. we're using the `$id` itself?

File: src/Symfony/Component/Translation/Tests/TranslatorCacheTest.php [hunk 3]
```diff
@@ -213,54 +213,6 @@ public function testDifferentCacheFilesAreUsedForDifferentSetsOfFallbackLocales(
         $this->assertEquals('bar', $translator->trans('bar'));
     }
 
-    public function testPrimaryAndFallbackCataloguesContainTheSameMessagesRegardlessOfCaching()
```
Comment (aitboudad) [inline src/Symfony/Component/Translation/Tests/TranslatorCacheTest.php]: it shoud be removed in `3.0` rename it into `testLegacyPrimaryA..` instead and add 
```php
$this->iniSet('error_reporting', -1 & ~E_USER_DEPRECATED);
```

File: src/Symfony/Component/Translation/Tests/TranslatorTest.php [hunk 4]
```diff
@@ -75,40 +74,6 @@ public function testSetValidLocale($locale)
         $this->assertEquals($locale, $translator->getLocale());
     }
 
-    public function testGetCatalogue()
```
Comment (aitboudad) [inline src/Symfony/Component/Translation/Tests/TranslatorTest.php]: same here

File: src/Symfony/Component/Translation/FallbackLocaleAwareInterface.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * An implementing class handles multiple locales and possible MessageCatalogues
+ * to find appropriate translations with some kind of fallback strategy.
+ *
+ * Through this interface it is possible to query which locale the translation
+ * for a given message id and domain will be taken from.
+ *
+ * @author Matthias Pigulla <mp@webfactory.de>
+ */
+interface FallbackLocaleAwareInterface
```
Comment (aitboudad) [inline src/Symfony/Component/Translation/FallbackLocaleAwareInterface.php]: `TranslatorLocaleAwareInterface` ?

File: src/Symfony/Component/Translation/FallbackLocaleAwareInterface.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    /**
+     * Determines the locale that the translation will ultimately be taken from.
+     *
+     * @param string      $id     The message id (may also be an object that can be cast to string)
+     * @param string|null $domain The domain for the message or null to use the default
+     * @param string|null $locale The locale or null to use the default
+     *
+     * @return string|null The locale of the best available translation or null for unknown messages.
+     */
+    public function resolveLocale($id, $domain = null, $locale = null);
```
Comment (aitboudad) [inline src/Symfony/Component/Translation/FallbackLocaleAwareInterface.php]: `getDefinesLocale` ?

File: src/Symfony/Component/Translation/Translator.php [hunk 7]
```diff
@@ -241,36 +241,52 @@ public function transChoice($id, $number, array $parameters = array(), $domain =
         }
 
         $id = (string) $id;
-        $catalogue = $this->getCatalogue($locale);
+        $bestLocale = $this->resolveLocale($id, $domain, $locale);
+
+        if (null !== $bestLocale) {
+           return strtr($this->selector->choose($this->getCatalogueInternal($locale)->get($id, $domain), (int) $number, $bestLocale), $parameters);
```
Comment (aitboudad) [inline src/Symfony/Component/Translation/Translator.php]: ``$this->getCatalogueInternal($bestLocale)``

File: src/Symfony/Component/Translation/Translator.php (hunk 1 above)
Comment (aitboudad) [inline src/Symfony/Component/Translation/Translator.php]: can you add this note above.

File: src/Symfony/Component/Translation/Translator.php (hunk 7 above)
Comment (mpdude) [inline src/Symfony/Component/Translation/Translator.php]: No, that might trigger loading another chain of catalogues (with the one for `bestLocale` at the root). As long as we're still using the fallback chain, it should be fine like this.

File: src/Symfony/Component/Translation/Tests/TranslatorCacheTest.php (hunk 3 above)
Comment (mpdude) [inline src/Symfony/Component/Translation/Tests/TranslatorCacheTest.php]: ✓

File: src/Symfony/Component/Translation/Translator.php (hunk 7 above)
Comment (aitboudad) [inline src/Symfony/Component/Translation/Translator.php]: right

File: src/Symfony/Component/Translation/Tests/TranslatorTest.php (hunk 4 above)
Comment (mpdude) [inline src/Symfony/Component/Translation/Tests/TranslatorTest.php]: ✓

**Comment by mpdude** [comment]
Thanks for the quick review @aitboudad!

File: src/Symfony/Component/Translation/DataCollectorTranslator.php [hunk 8]
```diff
@@ -35,8 +35,8 @@ class DataCollectorTranslator implements TranslatorInterface, TranslatorBagInter
      */
     public function __construct(TranslatorInterface $translator)
     {
-        if (!$translator instanceof TranslatorBagInterface) {
-            throw new \InvalidArgumentException(sprintf('The Translator "%s" must implement TranslatorInterface and TranslatorBagInterface.', get_class($translator)));
+        if (!($translator instanceof TranslatorBagInterface && $translator instanceof FallbackLocaleAwareInterface)) {
```
Comment (stof) [inline src/Symfony/Component/Translation/DataCollectorTranslator.php]: requiring it to implement both is actually a BC break as not implementing FallbackLocaleAwareInterface was allowed previously. You are restricting the supported argument, and the BC promise only allows to widen it

File: src/Symfony/Component/Translation/DataCollectorTranslator.php [hunk 9]
```diff
@@ -120,25 +128,25 @@ private function collectMessage($locale, $domain, $id, $translation)
...
-                }
-
-                $fallbackCatalogue = $fallbackCatalogue->getFallBackCatalogue();
-            }
-        } else {
-            $state = self::MESSAGE_MISSING;
+        $resolvedLocale = $this->translator->resolveLocale($id, $domain, $locale);
+
+        switch (true) {
+            case $resolvedLocale == $locale:
```
Comment (stof) [inline src/Symfony/Component/Translation/DataCollectorTranslator.php]: ``===``

File: src/Symfony/Component/Translation/LoggingTranslator.php [hunk 10]
```diff
@@ -34,8 +34,8 @@ class LoggingTranslator implements TranslatorInterface, TranslatorBagInterface
      */
     public function __construct(TranslatorInterface $translator, LoggerInterface $logger)
     {
-        if (!$translator instanceof TranslatorBagInterface) {
-            throw new \InvalidArgumentException(sprintf('The Translator "%s" must implement TranslatorInterface and TranslatorBagInterface.', get_class($translator)));
+        if (!($translator instanceof TranslatorBagInterface && $translator instanceof FallbackLocaleAwareInterface)) {
```
Comment (stof) [inline src/Symfony/Component/Translation/LoggingTranslator.php]: same issue here

File: src/Symfony/Component/Translation/LoggingTranslator.php [hunk 11]
```diff
@@ -113,16 +121,20 @@ private function log($id, $domain, $locale)
...
 
-        if ($catalogue->has($id, $domain)) {
-            $this->logger->debug('Translation use fallback catalogue.', array('id' => $id, 'domain' => $domain, 'locale' => $catalogue->getLocale()));
+        $id = (string) $id;
+
+        $resolvedLocale = $this->translator->resolveLocale($id, $domain, $locale);
+
+        if ($locale === $resolvedLocale) {
+            return;
+        } elseif ($resolvedLocale === null) {
```
Comment (stof) [inline src/Symfony/Component/Translation/LoggingTranslator.php]: ``if`` can be used rather than ``elseif`` as the ``if`` returns

File: src/Symfony/Component/Translation/Translator.php [hunk 12]
```diff
@@ -241,36 +241,52 @@ public function transChoice($id, $number, array $parameters = array(), $domain =
...
-            $locale = $this->getLocale();
-        } else {
-            $this->assertValidLocale($locale);
-        }
-
-        if (!isset($this->catalogues[$locale])) {
-            $this->loadCatalogue($locale);
-        }
+        trigger_error(
+            'The Translator will no longer implement TranslatorBagInterface in 3.0. If you want to find the locale actually used for a translation, use the resolveLocale() method instead.',
```
Comment (stof) [inline src/Symfony/Component/Translation/Translator.php]: You should use FQCNs in deprecation messages

File: src/Symfony/Component/Translation/Translator.php [hunk 13]
```diff
@@ -506,4 +522,26 @@ private function getConfigCacheFactory()
 
         return $this->configCacheFactory;
     }
+
+    /**
+     * Could be renamed back to getCatalogue() and made private once we drop TranslatorBagInterface in 3.0.
```
Comment (stof) [inline src/Symfony/Component/Translation/Translator.php]: the message is confusing. It is already private

File: src/Symfony/Component/Translation/DataCollectorTranslator.php (hunk 8 above)
Comment (stof) [inline src/Symfony/Component/Translation/DataCollectorTranslator.php]: the right solution would be to provide a Translator implementation which would implement ``FallbackLocaleAwareInterface`` on top of ``TranslatorBagInterface`` and wrap the translator into this implementation when it implements TranslatorBagInterface but not FallbackLocaleAwareInterface. This would make things BC, avoid duplication between the DataCollectorTranslator and the LoggingTranslator and not have performance penalty for the normal case (given that the default translator implements the new interface)

**Comment by mpdude** [comment]
@stof Thanks for your remarks. Does that mean that you basically support this idea?

**Comment by stof** [comment]
@mpdude I just thought about another case where TranslatorBagInterface could be useful: https://github.com/willdurand/BazingaJsTranslationBundle/ is exposing translation catalogues to the JS code. Currently, it does that by reimplementing the loading itself (it gets all translation loaders from Symfony and gets all resources as well, all of which is done by the Translator) because TranslatorBagInterface was not available at that time. What would be the clean way toimplement such your case in your new architecture (duplicating the loading is not the clean way in the 2.6 architecture IMO, but it was probably the only way in the 2.3 architecture, which is why it is used)

**Comment by aitboudad** [comment]
well I think we need to keep `getCatalogue` and the optimize caching should be done by optimizing `MessageCatalogue` see #14526

**Comment by mpdude** [comment]
@stof In the case of messages available in both the primary and in fallback catalogues, which set of messages do you think this bundle would need?

The problem is that if we burden `Translator` with general tasks like providing complete `MessageCatalogue` chains, it gets harder to optimize the data structures used internally for the "find best translation available" use case (the `trans()` and `transChoice()` methods). 

We could, of course, internally use optimized structures for these methods plus a "complete" set of `MessageCatalogue` chains for the `TranslatorBagInterface`-based use cases, but that would clearly be a nightmare to maintain and is not really a good example of applying the SRP.

So, what if we had two different classes/services (in the case of the full-stack framework):

The first one is where the resources and loaders are registered. You tell it your preferred order of locales, it will return a full-fledged `MessageCatalogue` chain (expensive operation!).

The second one is the `Translator`. It will initially obtain the `MessageCatalogue` from the first class, but then optimize/tweak things as it sees fit for the `trans()` and `transChoice()` methods. These optimized structures can be kept in the cache and will not be exposed to the outside world.

**Comment by stof** [comment]
@mpdude This bundle would need anything needed to be able to reimplement ``trans`` and ``transChoice`` in JS. So it would need the full catalogues for any active locales (it currently supports activating multiple locales, in case the locale choice need to be done on the JS side, or to keep a single translation file for all locales to simplify the loading of the JS), but only fallback messages might be enough for other locales.

And I agree that delegating all the loading to a separate class might be a better architecture. It could also simplify a bunch of things:

- lazy-loading of loaders could be implemented by implementing this interface in a lazy way (or relying of the DI feature to build the lazy implementation) instead of extending the Translator in FrameworkBundle, which would also make the lazy-loading available for people using DI+Translator outside Symfony
- caching of catalogues could be implemented by composition instead of being in the main class dealing with loaders (which does not forbid caching the internal structures of the Translator class, which is a separate topic).

This is something I already thought about in the past. The hard part is the Translator class is currently designed to be extendable through inheritance. Extracting this logic into separate collaborators in a BC way will likely be very hard.

**Comment by aitboudad** [comment]
@stof FYI I work on delegating all the loading :)

**Comment by stof** [comment]
@aitboudad have you found a way to keep BC for people overwriting protected methods of the Translator to customize the Translator through its current extension points ?

**Comment by aitboudad** [comment]
@stof see an initial work for [DelegationLoader](https://gist.github.com/aitboudad/6eed3ff1187bd8930520) I'll complete it soon, I think we need to deprecate `Symfony\Bundle\FrameworkBundle\Translation\Translator`.
To avoid BC we will create a new translator service and deprecate the default one.

**Comment by mpdude** [comment]
@stof I don't know if that's feasible. The Translator has some protected methods that deal with the process of loading the translations, so we can probably not move that responsibility out into another class.

**Comment by aitboudad** [comment]
@mpdude why we can't probably not move that responsibility out into another class ? until now 
I have only one issue about `initializeCatalogue`, I'll see how to deal with it.

@stof should we move handling fallback catalogues into `DelegationLoader` or by composition instead ?

**Comment by mpdude** [comment]
@stof but we can drop the protected methods in 3.0, right? What's the right way to deprecate them?

And we probably need a migration path as of 2.8. Does that mean we need to keep the method contracts as is, at best moving them to different classes? Or are we free to assign responsibilities differently to methods in 3.0?

**Comment by fabpot** [comment]
What is the status of this PR? ping @aitboudad

**Comment by mpdude** [comment]
I rebased this branch and fixed the various issues from the first review by @aitboudad and @stof.

This won't take us to the ultimate goal of a better and more efficient `Translator` (#13948) because we'd probably also need to factor loading out of the `Translator` and prohibit inheritance (maybe #14671 or #14622?)

Anyway, this improves things because we deprecate an interface that definetly is a show-stopper.

File: src/Symfony/Component/Translation/DataCollectorTranslator.php [hunk 14]
```diff
@@ -39,6 +44,12 @@ public function __construct(TranslatorInterface $translator)
             throw new \InvalidArgumentException(sprintf('The Translator "%s" must implement TranslatorInterface and TranslatorBagInterface.', get_class($translator)));
```
Comment (mpdude) [inline src/Symfony/Component/Translation/DataCollectorTranslator.php]: Need to fix that. The interface is deprecated and must not be required.

File: src/Symfony/Component/Translation/LoggingTranslator.php [hunk 15]
```diff
@@ -38,6 +43,12 @@ public function __construct(TranslatorInterface $translator, LoggerInterface $lo
             throw new \InvalidArgumentException(sprintf('The Translator "%s" must implement TranslatorInterface and TranslatorBagInterface.', get_class($translator)));
```
Comment (mpdude) [inline src/Symfony/Component/Translation/LoggingTranslator.php]: Same here

**Comment by mpdude** [comment]
Failed test is known to be transient (#15617).

**Comment by aitboudad** [comment]
@fabpot ll try to finish #14622 today as I think it's more efficient than this one(ideally we should have something like https://github.com/openl10n/openl10n-translation/tree/master/src/Openl10n/Component/Translation/MessageCatalogue)

**Comment by mpdude** [comment]
@symfony/deciders I would like to ask for your vote on this. 

I believe that removing the TranslatorBagInterface is one important prerequisite for more efficient caching and/or loading inside the Translator because it improves encapsulation.

With the current state of this PR, one important thing to note is that we leave clients without a way to retrieve the message catalogue from the translator, a feature added in 2.6 "by accident" (it would not have been necessary for the feature we added back then).

To provide an alternate means of retrieving the catalogue is an on-going effort in #14622 and #14671, but with the internal structure of `Translator` (different code paths, mixing several concerns, `protected` methods and `@api`) I am unsure if we can make it.

So, the decision you need to make is whether the improvement this PR brings is worth this possible annoyance for users of the `TranslatorBagInterface` in case that no other replacement for it can be provided.

Note that @stof [refers](https://github.com/symfony/symfony/pull/14530#issuecomment-98710622) to one third-party bundle that works *without* this interface by implementing the loading itself.

(The failing test is for PHP 7 is completely unrelated.)

**Comment by nicolas-grekas** [comment]
I think we don't need this one: optimized caching is already dealt with on PHP7, where static arrays and static strings are far superior optimizing strategies. Thus, no need to do anything more on our side to me.

**Comment by jakzal** [comment]
@mpdude any thoughts on what Nicolas said?

**Comment by nicolas-grekas** [comment]
Closing as explained. Of course, please reopen if you don't agree.

**Comment by mpdude** [comment]
I opened this a while ago when I was working on a project where we had issues due to a large number of languages supported and comprehensive catalogs. 

We found other ways to work around this and I cannot tell whether the problem would still appear on PHP7.

So, closing this is fine for me. Will reopen it or make a new case should I run into it again.
